packages/compile-solidity/test/compileWithPragmaAnalysis.js

```diff
@@ -149,6 +144,25 @@ describe("compileWithPragmaAnalysis", function () {
   });
 
   describe("when there is a semver expression error", function () {
+    it("throws an error when a 'direct source' lacks a pragma expression", async function () {
+      try {
+        await compileWithPragmaAnalysis({
+          options: config,
+          paths: [path.join(__dirname, "sources", "badSources", "NoPragma.sol")]
+        });
+        assert.fail("The function should have thrown.");
+      } catch (error) {
+        const expectedMessage = "Could not find a valid pragma expression";
+        if (error.message.includes(expectedMessage)) {
+          return "all good";
+        }
+        assert.fail(
```

Comment:
With this and the other unexpected error fails you added, I think it probably makes more sense to directly rethrow the unexpected error rather than using `assert.fail`; that way we get better stacktrace information.  (It's worth noting that doing things this way, we'll actually never get a stacktrace pointing to the `assert.fail` on line 153... because it'll be caught by the `catch` on line 154, and it's this `assert.fail` on line 159 that the stacktrace will actually point to!  Just rethrowing instead would avoid this problem, and other similar problems.)

Reply:
For all of these I thought it was more proper to use mocha mechanics to determine whether the tests failed or not. In other words, it seems slightly janky to throw an error in the test itself to make it fail as it is slightly less direct. But I opted to print the error in the "fail reason string" in order to keep the reason. You still think it would be better to throw? If so I don't really feel strongly enough about this to argue more :)

Reply:
I went ahead and just did a straight up assertion that the message was in the error and printed the error in the reason string. Let me know what you think of that.

Reply:
The new version seems to work by essentially the same mechanics as the old version?  It still has the same problems.  I'm not seeing any substantial difference here.

>  In other words, it seems slightly janky to throw an error in the test itself to make it fail as it is slightly less direct.

Well, it's not throwing a *new* error, it's *rethrowing* an existing error.  That's useful to preserve information; doing an assertion is actually throwing a new error and will obscure the source!  As I mentioned, currently one of your assertions will *never* cause the ultimate error because it just gets caught... that's confusing!